reco_utils/dataset/pandas_df_utils.py

```diff
@@ -402,3 +403,74 @@ def has_same_base_dtype(df_1, df_2, columns=None):
             result = False
 
     return result
+
+
+class PandasHash:
+    """Wrapper class to allow pandas DataFrames to be hashable"""
+
+    # reserve space just for a single DataFrame
+    __slots__ = 'df'
+
+    def __init__(self, df):
+        """Initialize class
+        Args:
+            df (pd.DataFrame): DataFrame
+        """
+        self.df = df
+
+    def __eq__(self, other):
+        """Overwrite equality comparison
+        Args:
+            other (pd.DataFrame): DataFrame to compare
+
+        Returns:
+            bool: whether other DataFrame is the same as this DataFrame
+        """
+        return hash(self) == hash(other)
+
+    def __hash__(self):
+        """Overwrite hash operator for use with DataFrames
+
+        Returns:
+            int: hashed value of DataFrame
+        """
+        return hash(tuple(pd.util.hash_pandas_object(self.df)) + tuple(self.df.columns))
+
+
+def lru_cache_df(maxsize, typed=False):
+    """Least-recently-used cache decorator
+
+    Args:
+        maxsize (int|None): max size of cache, if set to None cache is boundless
+        typed (bool): arguments of different types are cached separately
+    """
+
+    def to_pandas_hash(val):
+        """Return PandaHash object if input is a DataFrame otherwise return input unchanged"""
+        return PandasHash(val) if isinstance(val, pd.DataFrame) else val
+
+    def from_pandas_hash(val):
+        """Extract DataFrame if input is PandaHash object otherwise return input unchanged"""
+        return val.df if isinstance(val, PandasHash) else val
+
+    def decorating_function(user_function):
+        @wraps(user_function)
+        def wrapper(*args, **kwargs):
+            # convert DataFrames in args and kwargs to PandaHash objects
+            args = tuple([to_pandas_hash(a) for a in args])
+            kwargs = dict([(k, to_pandas_hash(v)) for k, v in kwargs.items()])
```

Comment:
This looks awesome @gramhagen. Maybe use dict comprehension here rather than instantiating dict here

Reply:
ok, good idea